// valorant.js/dist/resources/Endpoints.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Endpoints = void 0;
var Endpoints;
(function (Endpoints) {
    Endpoints["Auth"] = "https://auth.riotgames.com";
    Endpoints["Entitlements"] = "https://entitlements.auth.riotgames.com";
    Endpoints["NaBase"] = "https://pd.na.a.pvp.net";
    Endpoints["EuBase"] = "https://pd.eu.a.pvp.net";
    Endpoints["ApBase"] = "https://pd.ap.a.pvp.net";
    Endpoints["KrBase"] = "https://pd.kr.a.pvp.net";
    Endpoints["NaShared"] = "https://shared.na.a.pvp.net";
    Endpoints["EuShared"] = "https://shared.eu.a.pvp.net";
    Endpoints["ApShared"] = "https://shared.ap.a.pvp.net";
    Endpoints["KrShared"] = "https://shared.kr.a.pvp.net";
    Endpoints["NaParty"] = "https://glz-na-1.na.a.pvp.net";
    Endpoints["EuParty"] = "https://glz-eu-1.eu.a.pvp.net";
    Endpoints["ApParty"] = "https://glz-ap-1.ap.a.pvp.net";
    Endpoints["KrParty"] = "https://glz-kr-1.kr.a.pvp.net";
})(Endpoints = exports.Endpoints || (exports.Endpoints = {}));
