// valorant.js/dist/index.js
"use strict";
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, privateMap, value) {
    if (!privateMap.has(receiver)) {
        throw new TypeError("attempted to set private field on non-instance");
    }
    privateMap.set(receiver, value);
    return value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, privateMap) {
    if (!privateMap.has(receiver)) {
        throw new TypeError("attempted to get private field on non-instance");
    }
    return privateMap.get(receiver);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _config;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Region = exports.Http = exports.RiotApiClient = void 0;
const ContentApi_1 = require("./services/ContentApi");
const MatchApi_1 = require("./services/MatchApi");
const PlayerApi_1 = require("./services/PlayerApi");
const StoreApi_1 = require("./services/StoreApi");
const PartyApi_1 = require("./services/PartyApi");
const Http_1 = require("./Http");
const axios_1 = __importDefault(require("axios"));
const Request_1 = require("./Request");
const Endpoints_1 = require("./resources/Endpoints");
const Exceptions_1 = require("./models/Exceptions");
const tough_cookie_1 = require("tough-cookie");
const axios_cookiejar_support_1 = __importDefault(require("axios-cookiejar-support"));
class RiotApiClient {
    /**
     * - Initiates the base client
     * @param config Config for the lib
     */
    constructor(config) {
        _config.set(this, void 0);
        if (!(config.region instanceof Region))
            throw new Error("'Config.region' must be type of 'Region'.");
        this.http = new Http();
        __classPrivateFieldSet(this, _config, config);
        this.region = config.region;
        this.buildServices();
    }
    /**
     * - Logins into your account
     */
    async login() {
        // login and setup some stuff
        this.auth = {};
        this.auth.accessToken = await this.playerApi.getAccessToken(__classPrivateFieldGet(this, _config).username, __classPrivateFieldGet(this, _config).password);
        this.auth.rsoToken = await this.playerApi.getRsoToken(this.auth.accessToken);
        this.buildServices();
        // get user
        const userInfo = await this.playerApi.getInfo();
        if (userInfo.sub == "")
            throw new Error("Account ID was empty. Please start the game atleast once!");
        this.user = (await this.playerApi.getAccountById([userInfo.sub]))[0];
        // finish stuff
        this.clientVersion = await this.getClientVersion();
        this.buildServices();
        return this;
    }
    /**
     * - Gets the current client version
     */
    async getClientVersion() {
        try {
            const data = (await axios_1.default({
                method: "GET",
                url: "https://valorant-api.com/v1/version"
            })).data.data;
            const branch = data.branch;
            const build = data.buildVersion;
            const versionNum = data.version.split(".").pop();
            return branch + "-shipping-" + build + "-" + versionNum;
        }
        catch (e) {
            console.error(e);
        }
    }
    /**
     * - Initiates services
     * @warning You probably shouldn't call this method
     */
    buildServices() {
        this.storeApi = new StoreApi_1.StoreApi(this);
        this.partyApi = new PartyApi_1.PartyApi(this);
        this.playerApi = new PlayerApi_1.PlayerApi(this);
        this.contentApi = new ContentApi_1.ContentApi(this);
        this.matchApi = new MatchApi_1.MatchApi(this);
        this.http = new Http(this.auth, this.clientVersion);
    }
}
exports.RiotApiClient = RiotApiClient;
_config = new WeakMap();
/**
 * - Client platform id
 */
RiotApiClient.XRiotClientPlatform = "ew0KCSJwbGF0Zm9ybVR5cGUiOiAiUEMiLA0KCSJwbGF0Zm9ybU9TIjogIldpbmRvd3MiLA0KCSJwbGF0Zm9ybU9TVmVyc2lvbiI6ICIxMC4wLjE5MDQyLjEuMjU2LjY0Yml0IiwNCgkicGxhdGZvcm1DaGlwc2V0IjogIlVua25vd24iDQp9";
axios_cookiejar_support_1.default(axios_1.default);
class Http extends Http_1.AbstractHttp {
    constructor(authorization, version) {
        super();
        this.auth = null;
        this.version = null;
        this.jar = new tough_cookie_1.CookieJar();
        this.auth = authorization;
        this.version = version;
    }
    /**
     * - Sends a request
     * @param request Request to send
     * @throws {ApiClientException}
     */
    async sendRequest(request) {
        try {
            const modifiedReq = Request_1.RequestBuilder.fromRequest(request);
            if (this.auth != null && this.auth.accessToken != null) {
                modifiedReq.addHeader("Authorization", `${this.auth.accessToken.token_type} ${this.auth.accessToken.access_token}`);
                if (this.auth.rsoToken != null)
                    modifiedReq.addHeader("X-Riot-Entitlements-JWT", this.auth.rsoToken.entitlements_token);
            }
            if (this.version != null)
                modifiedReq.addHeader("X-Riot-ClientVersion", this.version);
            modifiedReq.addHeader("X-Riot-ClientPlatform", RiotApiClient.XRiotClientPlatform);
            modifiedReq.setCookieJar(this.jar);
            return await axios_1.default(modifiedReq.build());
        }
        catch (e) {
            throw e.response
                ? new Exceptions_1.ApiClientException(e)
                : e;
        }
    }
}
exports.Http = Http;
class Region {
    constructor(baseUrl, sharedUrl, partyUrl, name) {
        this.BaseUrl = baseUrl;
        this.SharedUrl = sharedUrl;
        this.PartyUrl = partyUrl;
        this.Name = name;
    }
}
exports.Region = Region;
Region.EU = new Region(Endpoints_1.Endpoints.EuBase, Endpoints_1.Endpoints.EuShared, Endpoints_1.Endpoints.EuParty, "eu");
Region.NA = new Region(Endpoints_1.Endpoints.NaBase, Endpoints_1.Endpoints.NaShared, Endpoints_1.Endpoints.NaParty, "na");
Region.AP = new Region(Endpoints_1.Endpoints.ApBase, Endpoints_1.Endpoints.ApShared, Endpoints_1.Endpoints.ApParty, "ap");
Region.KR = new Region(Endpoints_1.Endpoints.KrBase, Endpoints_1.Endpoints.KrShared, Endpoints_1.Endpoints.KrParty, "kr");
