// @valapi/valorant-api.com/dist/service/Seasons.js
"use strict";
//import
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Seasons = void 0;
//class
class Seasons {
    constructor(AxiosClient, language) {
        this.AxiosClient = AxiosClient;
        this.language = language;
    }
    //service
    get() {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.AxiosClient.request('/seasons' + `?language=${this.language}`);
        });
    }
    getCompetitiveSeasons() {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.AxiosClient.request('/seasons/competitive');
        });
    }
    getByUuid(uuid) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.AxiosClient.request(`/seasons/${uuid}` + `?language=${this.language}`);
        });
    }
    getCompetitiveSeasonByUuid(uuid) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.AxiosClient.request(`/seasons/competitive/${uuid}`);
        });
    }
}
exports.Seasons = Seasons;
